db/buffered_insert.go

```diff
@@ -0,0 +1,218 @@
+		catcher.Add(errors.New("must specify a collection name"))
+	}
+
+	if o.Count < 2 {
+		catcher.Add(errors.New("must specify a count of at least 2"))
+	}
+
+	if o.Duration < 10*time.Millisecond {
+		catcher.Add(errors.New("must specify a duration of at least 10ms"))
+	}
+
+	return catcher.Resolve()
+}
+
+type anserBufInsertsImpl struct {
+	opts    BufferedInsertOptions
+	db      Database
+	cancel  context.CancelFunc
+	docs    chan interface{}
+	flusher chan chan error
+	closer  chan chan error
+	err     chan error
+}
+
+// Constructs a buffered insert handler, wrapping an mgo-like
+// interface for a database dervied from anser's database
+// interface. You must specify the collection that the
+// inserts target in the options. You can construct the same instance
+// using an actual mgo session with the NewBufferedSession
+func NewBufferedInserter(ctx context.Context, db Database, opts BufferedInsertOptions) (BufferedInserter, error) {
+	if err := opts.Validate(); err != nil {
+		return nil, errors.Wrap(err, "cannot construct buffered insert handler")
+	}
+
+	bi := &anserBufInsertsImpl{
+		docs:    make(chan interface{}),
+		err:     make(chan error),
+		flusher: make(chan chan error),
+		closer:  make(chan chan error),
+		db:      db,
+		opts:    opts,
+	}
+	ctx, bi.cancel = context.WithCancel(ctx)
+
+	go bi.start(ctx)
+
+	return bi, nil
+}
+
+func NewBufferedSessionInserter(ctx context.Context, s *mgo.Session, opts BufferedInsertOptions) (BufferedInserter, error) {
+	if opts.DB == "" {
+		return nil, errors.New("must specify a database name when constructing a buffered insert handler")
+	}
+
+	if s == nil {
+		return nil, errors.New("must specify a non-nil database session")
+	}
+
+	session := WrapSession(s)
+
+	return NewBufferedInserter(ctx, session.DB(opts.DB), opts)
+}
+
+func (bi *anserBufInsertsImpl) start(ctx context.Context) {
+	timer := time.NewTimer(bi.opts.Duration)
+	defer timer.Stop()
+	buf := []interface{}{}
+	catcher := grip.NewBasicCatcher()
+
+bufferLoop:
+	for {
+		select {
+		case <-ctx.Done():
+			if len(buf) > 0 {
+				catcher.Add(errors.Errorf("buffered insert has %d pending inserts", len(buf)))
+			}
+
+			bi.err <- catcher.Resolve()
+			break bufferLoop
+		case <-timer.C:
+			if len(buf) > 0 {
+				err := bi.db.C(bi.opts.Collection).Insert(buf...)
+				catcher.Add(err)
+				if err == nil {
+					buf = []interface{}{}
+				}
+			}
+
+			timer.Reset(bi.opts.Duration)
+		case doc := <-bi.docs:
+			buf = append(buf, doc)
+
+			if len(buf) >= bi.opts.Count {
+				err := bi.db.C(bi.opts.Collection).Insert(buf...)
+				catcher.Add(err)
+				if err == nil {
+					buf = []interface{}{}
+				}
+			}
+		case f := <-bi.flusher:
+			select {
+			case last := <-bi.docs:
+				buf = append(buf, last)
+			default:
+			}
+
+			if len(buf) > 0 {
+				err := bi.db.C(bi.opts.Collection).Insert(buf...)
+				catcher.Add(err)
+				if err == nil {
+					buf = []interface{}{}
+				}
+				f <- err
+				timer.Reset(bi.opts.Duration)
```

Comment:
I think you may need to
```
if !timer.Stop() {
	<-t.C
}
timer.Reset(d)
```
The docs for timers mention it is racy otherwise.

Reply:
done